reference/distributed/partition_helpers.hpp

```diff
@@ -48,6 +51,51 @@ LocalIndexType map_to_local(
 }
 
 
+template <typename LocalIndexType, typename GlobalIndexType>
+size_type find_local_range(
+    LocalIndexType idx, size_type part_id,
+    device_partition<const LocalIndexType, const GlobalIndexType> partition,
+    const size_type local_range_id_hint = 0)
+{
+    const auto& ranges_by_part = partition.ranges_by_part;
+    auto local_ranges = ranges_by_part.get_segment(part_id);
+    auto local_range_size =
+        static_cast<size_type>(local_ranges.end - local_ranges.begin);
+
+    auto range_starting_indices = partition.starting_indices_begin;
+    if (range_starting_indices[local_ranges.begin[local_range_id_hint]] <=
+            idx &&
+        (local_range_id_hint == local_range_size - 1 ||
+         range_starting_indices[local_ranges.begin[local_range_id_hint + 1]] >
+             idx)) {
+        return local_range_id_hint;
+    }
+
+    auto it = std::upper_bound(
+        local_ranges.begin, local_ranges.end, idx,
+        [range_starting_indices](const auto value, const auto rid) {
+            return value < range_starting_indices[rid];
+        });
+    auto local_range_id = std::distance(local_ranges.begin, it) - 1;
+    return local_range_id;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+GlobalIndexType map_to_global(
+    LocalIndexType idx,
+    device_partition<const LocalIndexType, const GlobalIndexType> partition,
+    size_type range_id)
+{
+    assert(range_id <
+           std::distance(partition.offsets_begin, partition.offsets_end) - 1);
```

Comment:
```suggestion
    assert(range_id <
           std::distance(partition.offsets_begin, partition.offsets_end));
```
do you need -1?

Reply:
Yes, the offsets array has `#range-ids + 1` entries, since it's an offset array. For an empty partition, `offsets = {0}`, so the begin and end pointers are different.